--- mutt-1.3.18/send.c	Thu Apr 26 15:17:49 2001
+++ mutt-1.3.18-p/send.c	Wed May 30 13:30:10 2001
@@ -630,8 +630,10 @@
   else
   {
     env->references = mutt_make_references (curenv);
-    env->in_reply_to = mutt_new_list ();
-    env->in_reply_to->data = safe_strdup (curenv->message_id);
+    if (curenv->message_id) {
+      env->in_reply_to = mutt_new_list ();
+      env->in_reply_to->data = safe_strdup (curenv->message_id);
+    }
   }
 }
 
