diff -urN mutt-1.5.19/doc/Muttrc mutt-1.5.19.conf/doc/Muttrc
--- mutt-1.5.19/doc/Muttrc	2009-04-13 19:07:38.103582891 +0200
+++ mutt-1.5.19.conf/doc/Muttrc	2009-04-13 18:29:07.652668002 +0200
@@ -1033,6 +1033,7 @@
 # 
 # 
 # set forward_format="[%a: %s]"
+set forward_format="Fwd: %s"
 #
 # Name: forward_format
 # Type: string
@@ -1203,6 +1204,7 @@
 # 
 # 
 # set history=10
+set history=100
 #
 # Name: history
 # Type: number
@@ -1775,6 +1777,7 @@
 # 
 # 
 # set mark_old=yes
+set mark_old=no
 #
 # Name: mark_old
 # Type: boolean
@@ -1947,6 +1950,7 @@
 # 
 # 
 # set mime_forward=no
+set mime_forward=ask-no
 #
 # Name: mime_forward
 # Type: quadoption
@@ -2115,6 +2119,7 @@
 # 
 # 
 # set pager_context=0
+set pager_context=1
 #
 # Name: pager_context
 # Type: number
@@ -2141,6 +2146,7 @@
 # 
 # 
 # set pager_index_lines=0
+set pager_index_lines=6
 #
 # Name: pager_index_lines
 # Type: number
@@ -3208,6 +3214,7 @@
 # 
 # 
 # set ssl_ca_certificates_file=""
+set ssl_ca_certificates_file="/etc/certs/ca-certificates.crt"
 #
 # Name: ssl_ca_certificates_file
 # Type: path
@@ -3546,6 +3553,7 @@
 # 
 # 
 # set quit=yes
+set quit=ask-yes
 #
 # Name: quit
 # Type: quadoption
@@ -3680,6 +3688,7 @@
 # 
 # 
 # set reply_regexp="^(re([\\[0-9\\]+])*|aw):[ \t]*"
+set reply_regexp="^(([oO][dD][pP]|[aA][wW]|[rR][eE]([\\[0-9\\]+])*:|[Oo][dD][pP]:|[aA][wW]:|[rR][eE]:)[ \t]+)*"
 #
 # Name: reply_regexp
 # Type: regular expression
@@ -3915,6 +3924,7 @@
 # 
 # 
 # set send_charset="us-ascii:iso-8859-1:utf-8"
+set send_charset="us-ascii:utf-8"
 #
 # Name: send_charset
 # Type: string
@@ -4019,6 +4029,7 @@
 # 
 # 
 # set simple_search="~f %s | ~s %s"
+set simple_search="~f %s | ~C %s | ~s %s"
 #
 # Name: simple_search
 # Type: string
@@ -4126,6 +4137,7 @@
 # 
 # 
 # set sort=date
+set sort=threads
 #
 # Name: sort
 # Type: sort order
@@ -4436,6 +4448,7 @@
 # 
 # 
 # set thorough_search=no
+set thorough_search=yes
 #
 # Name: thorough_search
 # Type: boolean
@@ -4455,6 +4468,7 @@
 # 
 # 
 # set tilde=no
+set tilde=yes
 #
 # Name: tilde
 # Type: boolean
@@ -4772,3 +4786,4 @@
 # in an X11 terminal emulator.
 # 
 # 
+source mutt_source-muttrc.d|
