--- mutt-1.5.21/configure.ac.orig	2012-04-06 17:33:23.905605556 +0200
+++ mutt-1.5.21/configure.ac	2012-04-07 07:42:20.793977750 +0200
@ -985,20 +985,20 @@
                 bdbpfx="$bdbpfx $d/$v"
             done
         done
-        BDB_VERSIONS="db-4 db4 db-5 db5 db-5.2 db5.2 db52 db-5.1 db5.1 db51 db-5.0 db5.0 db50 db-4.8 db4.8 db48 db-4.7 db4.7 db47 db-4.6 db4.6 db46 db-4.5 db4.5 db45 db-4.4 db4.4 db44 db-4.3 db4.3 db43 db-4.2 db4.2 db42 db-4.1 db4.1 db41 db ''"
+        BDB_VERSIONS="db-5.3 db-5.2 db5.2 db52 db-5.1 db5.1 db51 db-5.0 db5.0 db50 db-5 db5 db-4.8 db4.8 db48 db-4.7 db4.7 db47 db-4.6 db4.6 db46 db-4.5 db4.5 db45 db-4.4 db4.4 db44 db-4.3 db4.3 db43 db-4.2 db4.2 db42 db-4.1 db4.1 db41 db-4 db4 db ''"
         AC_MSG_CHECKING([for BerkeleyDB > 4.0])
         for d in $bdbpfx; do
             BDB_INCLUDE_DIR=""
-            BDB_LIB_DIR=""
+            BDB_LIB_L=""
             for v in / $BDB_VERSIONS; do
                 if test -r "$d/include/$v/db.h"; then
                     BDB_INCLUDE_DIR="$d/include/$v"
-                    for bdblibdir in "$d/lib/$v" "$d/lib"; do
-                        test -d "$bdblibdir" || continue
-                        BDB_LIB_DIR="$bdblibdir"
+                    for bdblibdir in '' "$d/lib/$v" "$d/lib"; do
+                        test -z "$bdblibdir" -o -d "$bdblibdir" || continue
+                        BDB_LIB_L="${bdblibdir:+-L$bdblibdir}"
                         for l in `echo $BDB_VERSIONS`; do
                             CPPFLAGS="$OLDCPPFLAGS -I$BDB_INCLUDE_DIR"
-                            LIBS="$OLDLIBS -L$BDB_LIB_DIR -l$l"
+                            LIBS="$OLDLIBS $BDB_LIB_L -l$l"
                             AC_LINK_IFELSE([AC_LANG_PROGRAM([[
                                 #include <stdlib.h>
                                 #include <db.h>
@@ -1021,7 +1021,7 @@
         then
             AC_MSG_RESULT(yes)
             CPPFLAGS="$OLDCPPFLAGS -I$BDB_INCLUDE_DIR"
-            LIBS="$OLDLIBS -L$BDB_LIB_DIR -l$BDB_LIB"
+            LIBS="$OLDLIBS $BDB_LIB_L -l$BDB_LIB"
             AC_DEFINE(HAVE_DB4, 1, [Berkeley DB4 Support])
             db_found=bdb
         else
