--- mutt-1.4.2.1/configure.in.orig	2005-08-21 21:28:37.000000000 +0200
+++ mutt-1.4.2.1/configure.in	2005-08-21 21:30:40.000000000 +0200
@@ -42,9 +42,9 @@
 AC_C_INLINE
 AC_C_CONST
 
-AC_PATH_PROG(DBX, dbx, no)
-AC_PATH_PROG(GDB, gdb, no)
-AC_PATH_PROG(SDB, sdb, no)
+DBX=no
+GDB=/usr/bin/gdb
+SDB=no
 
 if test $GDB != no ; then
         DEBUGGER=$GDB
@@ -58,8 +58,7 @@
 
 AC_SUBST(DEBUGGER)
 
-ac_aux_path_sendmail=/usr/sbin:/usr/lib
-AC_PATH_PROG(SENDMAIL, sendmail, /usr/sbin/sendmail, $PATH:$ac_aux_path_sendmail)
+ac_cv_path_SENDMAIL=/usr/sbin/sendmail
 AC_DEFINE_UNQUOTED(SENDMAIL,"$ac_cv_path_SENDMAIL",[ Where to find sendmail on your system. ])
 
 OPS='$(srcdir)/OPS'
@@ -101,7 +100,7 @@
 AC_DEFINE_UNQUOTED(SUBVERSION,"$SUBVERSION",[ Is this the international version? ])
 AC_SUBST(SUBVERSION)
 
-AC_PATH_PROG(ISPELL, ispell, no)
+ISPELL=/usr/bin/ispell
 if test $ISPELL != no; then
         AC_DEFINE_UNQUOTED(ISPELL,"$ISPELL",[ Where to find ispell on your system. ])
 fi
