--- mutt-1.4.2.2/imap/auth_sasl.c.orig	2006-08-21 20:08:06.165971000 +0200
+++ mutt-1.4.2.2/imap/auth_sasl.c	2006-08-21 20:32:39.285971000 +0200
@@ -32,7 +32,7 @@
   sasl_conn_t* saslconn;
   sasl_interact_t* interaction = NULL;
   int rc, irc;
-  char buf[LONG_STRING];
+  char buf[HUGE_STRING];
   const char* mech;
   const char* pc = NULL;
   unsigned int len, olen;
@@ -107,7 +107,7 @@
     if (irc == IMAP_CMD_RESPOND)
     {
       if (sasl_decode64 (idata->cmd.buf+2, strlen (idata->cmd.buf+2), buf,
-			 LONG_STRING,&len) != SASL_OK)
+			 LONG_STRING-1, &len) != SASL_OK)
       {
 	dprint (1, (debugfile, "imap_auth_sasl: error base64-decoding server response.\n"));
 	goto bail;
--- mutt-1.4.2.2/mutt_sasl.c.orig	2006-08-21 20:08:06.205971000 +0200
+++ mutt-1.4.2.2/mutt_sasl.c	2006-08-21 20:31:54.415971000 +0200
@@ -383,7 +383,6 @@
 
   /* release sasl resources */
   sasl_dispose (&sasldata->saslconn);
-  FREE (&sasldata->buf);
   FREE (&sasldata);
 
   /* call underlying close */
@@ -415,7 +414,6 @@
   
   conn->sockdata = sasldata->sockdata;
 
-  FREE (&sasldata->buf);
   sasldata->bpos = 0;
   sasldata->blen = 0;
 
@@ -486,7 +484,6 @@
       }
 
       rc = (sasldata->write) (conn, pbuf, plen);
-      FREE (&pbuf);
       if (rc != plen)
 	goto fail;
 
