--- mutt-1.2.5/muttbug.sh.in.foo	Mon May 14 10:09:57 2001
+++ mutt-1.2.5/muttbug.sh.in	Mon May 14 10:12:41 2001
@@ -85,13 +85,8 @@
 exec > /dev/tty
 exec < /dev/tty
 
-SCRATCH=${TMPDIR-/tmp}/`basename $0`.`hostname`.$$
-
-mkdir ${SCRATCH} || \
-{ 
-	echo "`basename $0`: Can't create temporary directory." >& 2 ; 
-	exit 1 ; 
-}
+TMP=${TMPDIR-/tmp}/`basename $0`.`hostname`
+SCRATCH=`mktemp -d $TMP.XXXXXX || exit 1`
 
 trap "rm -r -f ${SCRATCH} ; trap '' 0 ; exit" 0 1 2
 
