diff -urN mutt-1.3.27.orig/browser.c mutt-1.3.27/browser.c
--- mutt-1.3.27.orig/browser.c	Tue Dec 18 16:12:48 2001
+++ mutt-1.3.27/browser.c	Mon Feb  4 00:40:48 2002
@@ -388,6 +388,10 @@
 
   while ((de = readdir (dp)) != NULL)
   {
+    /* It can happen because of broken VFAT driver. */
+    if (mutt_strcmp (de->d_name, "") == 0)
+      continue;
+
     if (mutt_strcmp (de->d_name, ".") == 0)
       continue;    /* we don't need . */
     
