diff -urNp -x '*.orig' mutt-2.0.6.org/doc/Muttrc.head mutt-2.0.6/doc/Muttrc.head
--- mutt-2.0.6.org/doc/Muttrc.head	2020-07-25 20:22:53.000000000 +0200
+++ mutt-2.0.6/doc/Muttrc.head	2021-04-18 19:23:39.000021331 +0200
@@ -23,7 +23,7 @@ macro index,pager,attach,compose \cb "\
 "call urlview to extract URLs out of a message"
 
 # Show documentation when pressing F1
-macro generic,pager <F1> "<shell-escape> less @docdir@/manual.txt<Enter>" "show Mutt documentation"
+macro generic,pager <F1> "<shell-escape> less @docdir@/manual.txt.gz<Enter>" "show Mutt documentation"
 
 # show the incoming mailboxes list (just like "mutt -y") and back when pressing "y"
 # note: these macros have been subsumed by the <browse-mailboxes> function.
