diff -urNp -x '*.orig' mutt-2.0.6.org/browser.c mutt-2.0.6/browser.c
--- mutt-2.0.6.org/browser.c	2020-12-01 04:05:21.000000000 +0100
+++ mutt-2.0.6/browser.c	2021-04-18 19:23:38.053342162 +0200
@@ -475,6 +475,10 @@ static int examine_directory (MUTTMENU *
 
   while ((de = readdir (dp)) != NULL)
   {
+    /* It can happen because of broken VFAT driver. */
+    if (mutt_strcmp (de->d_name, "") == 0)
+      continue;
+
     if (mutt_strcmp (de->d_name, ".") == 0)
       continue;    /* we don't need . */
 
