diff -urN mutt-1.3.17.org/configure.in mutt-1.3.17/configure.in
--- mutt-1.3.17.org/configure.in	Wed Apr  4 23:31:32 2001
+++ mutt-1.3.17/configure.in	Wed Apr  4 23:33:55 2001
@@ -321,41 +321,8 @@
 	fi
 	AC_DEFINE_UNQUOTED(MAILPATH, "$mutt_cv_mailpath")
 
-	AC_CACHE_CHECK(if $mutt_cv_mailpath is world writable, mutt_cv_worldwrite, [AC_TRY_RUN([#include <sys/types.h>
-#include <sys/stat.h>
-
-int main (int argc, char **argv)
-{
-	struct stat s;
-
-	stat ("$mutt_cv_mailpath", &s);
-	if (s.st_mode & S_IWOTH) exit (0);
-	exit (1);
-}], mutt_cv_worldwrite=yes, mutt_cv_worldwrite=no, mutt_cv_worldwrite=no)])
-
-	mutt_cv_setgid=no
-	if test $mutt_cv_worldwrite = yes; then
-		AC_DEFINE(USE_DOTLOCK)
-	else
-
-		AC_CACHE_CHECK(if $mutt_cv_mailpath is group writable, mutt_cv_groupwrite, [AC_TRY_RUN([#include <sys/types.h>
-#include <sys/stat.h>
-
-int main (int argc, char **argv)
-{
-	struct stat s;
-
-	stat ("$mutt_cv_mailpath", &s);
-	if (s.st_mode & S_IWGRP) exit (0);
-	exit (1);
-}], mutt_cv_groupwrite=yes, mutt_cv_groupwrite=no, mutt_cv_groupwrite=no)])
-
-		if test $mutt_cv_groupwrite = yes; then
-			AC_DEFINE(USE_DOTLOCK)
-			AC_DEFINE(USE_SETGID)
-			mutt_cv_setgid=yes
-		fi
-	fi
+	AC_DEFINE(USE_DOTLOCK)
+	AC_DEFINE(USE_SETGID)
 fi
 
 AC_ARG_ENABLE(external_dotlock, [  --enable-external-dotlock  Force use of an external dotlock program],
