--- mutt-2.2.13/contrib/smime.rc.orig	2023-08-18 05:03:18.000000000 +0200
+++ mutt-2.2.13/contrib/smime.rc	2024-10-27 16:30:29.327806343 +0100
@@ -47,13 +47,14 @@
 # unset smime_decrypt_use_default_key
 
 # Path to a file or directory with trusted certificates
-set smime_ca_location="~/.smime/ca-bundle.crt"
+# set smime_ca_location="~/.smime/ca-bundle.crt"
+set smime_ca_location="/etc/certs/ca-certificates.crt"
 
 # Path to where all known certificates go. (must exist!)
 set smime_certificates="~/.smime/certificates"
 
 # Path to where all private keys go. (must exist!)
-set smime_keys="~/.smime/keys"
+# set smime_keys="~/.smime/keys"
 
 # These are used to extract a certificate from a message.
 # First generate a PKCS#7 structure from the message.
