--- mutt-2.0.6/configure.ac.orig	2021-03-06 19:06:37.000000000 +0000
+++ mutt-2.0.6/configure.ac	2021-04-07 17:45:51.050794394 +0000
@@ -1180,16 +1180,16 @@
         AC_MSG_CHECKING([for BerkeleyDB > 4.0])
         for d in $bdbpfx; do
             BDB_INCLUDE_DIR=""
-            BDB_LIB_DIR=""
+            BDB_LIB_L=""
             for v in / $BDB_VERSIONS; do
                 if test -r "$d/include/$v/db.h"; then
                     BDB_INCLUDE_DIR="$d/include/$v"
-                    for bdblibdir in "$d/lib/$v" "$d/lib"; do
-                        test -d "$bdblibdir" || continue
-                        BDB_LIB_DIR="$bdblibdir"
+                    for bdblibdir in '' "$d/lib/$v" "$d/lib"; do
+			test -z "$bdblibdir" -o -d "$bdblibdir" || continue
+			BDB_LIB_L="${bdblibdir:+-L$bdblibdir}"
                         for l in `echo $BDB_VERSIONS`; do
                             CPPFLAGS="$OLDCPPFLAGS -I$BDB_INCLUDE_DIR"
-                            LIBS="$OLDLIBS -L$BDB_LIB_DIR -l$l"
+			    LIBS="$OLDLIBS $BDB_LIB_L -l$l"
                             AC_LINK_IFELSE([AC_LANG_PROGRAM([[
                                 #include <stdlib.h>
                                 #include <db.h>
@@ -1212,7 +1212,7 @@
         then
             AC_MSG_RESULT(yes)
             CPPFLAGS="$OLDCPPFLAGS -I$BDB_INCLUDE_DIR"
-            LIBS="$OLDLIBS -L$BDB_LIB_DIR -l$BDB_LIB"
+            LIBS="$OLDLIBS $BDB_LIB_L -l$BDB_LIB"
             AC_DEFINE(HAVE_DB4, 1, [Berkeley DB4 Support])
             db_found=bdb
         else
