--- mutt-0.95-us/configure.in.nosetgid	Sat Dec 12 04:22:04 1998
+++ mutt-0.95-us/configure.in	Sat Dec 12 22:31:34 1998
@@ -302,25 +302,6 @@
 	mutt_cv_setgid=no
 	if test $mutt_cv_worldwrite = yes; then
 		AC_DEFINE(USE_DOTLOCK)
-	else
-
-		AC_CACHE_CHECK(if $mutt_cv_mailpath is group writable, mutt_cv_groupwrite, [AC_TRY_RUN([#include <sys/types.h>
-#include <sys/stat.h>
-
-int main (int argc, char **argv)
-{
-	struct stat s;
-
-	stat ("$mutt_cv_mailpath", &s);
-	if (s.st_mode & S_IWGRP) exit (0);
-	exit (1);
-}], mutt_cv_groupwrite=yes, mutt_cv_groupwrite=no, mutt_cv_groupwrite=no)])
-
-		if test $mutt_cv_groupwrite = yes; then
-			AC_DEFINE(USE_DOTLOCK)
-			AC_DEFINE(USE_SETGID)
-			mutt_cv_setgid=yes
-		fi
 	fi
 fi
 
