diff -pruN1 mutt-1.3.25.trash_folder/OPS mutt-1.3.25.purge_command/OPS
--- mutt-1.3.25.trash_folder/OPS	Tue Sep 11 12:50:50 2001
+++ mutt-1.3.25.purge_command/OPS	Wed Jan  9 20:40:27 2002
@@ -136,2 +136,3 @@ OP_PREV_PAGE "move to the previous page"
 OP_PRINT "print the current entry"
+OP_PURGE_MESSAGE "really delete the current entry, bypassing the trash folder"
 OP_QUERY "query external program for addresses"
diff -pruN1 mutt-1.3.25.trash_folder/curs_main.c mutt-1.3.25.purge_command/curs_main.c
--- mutt-1.3.25.trash_folder/curs_main.c	Thu Nov  8 09:56:02 2001
+++ mutt-1.3.25.purge_command/curs_main.c	Wed Jan  9 20:40:27 2002
@@ -1615,2 +1615,3 @@ int mutt_index_menu (void)
 
+      case OP_PURGE_MESSAGE:
       case OP_DELETE:
@@ -1623,2 +1624,4 @@ int mutt_index_menu (void)
 	  mutt_tag_set_flag (M_DELETE, 1);
+	  if (op == OP_PURGE_MESSAGE)
+	    mutt_tag_set_flag (M_PURGED, 1);
 	  if (option (OPTDELETEUNTAG))
@@ -1630,2 +1633,4 @@ int mutt_index_menu (void)
 	  mutt_set_flag (Context, CURHDR, M_DELETE, 1);
+	  if (op == OP_PURGE_MESSAGE)
+	    mutt_set_flag (Context, CURHDR, M_PURGED, 1);
 	  if (option (OPTDELETEUNTAG))
@@ -1908,2 +1913,3 @@ int mutt_index_menu (void)
 	  mutt_tag_set_flag (M_DELETE, 0);
+	  mutt_tag_set_flag (M_PURGED, 0);
 	  menu->redraw = REDRAW_INDEX;
@@ -1913,2 +1919,3 @@ int mutt_index_menu (void)
 	  mutt_set_flag (Context, CURHDR, M_DELETE, 0);
+	  mutt_set_flag (Context, CURHDR, M_PURGED, 0);
 	  if (option (OPTRESOLVE) && menu->current < Context->vcount - 1)
@@ -1932,5 +1939,7 @@ int mutt_index_menu (void)
 	rc = mutt_thread_set_flag (CURHDR, M_DELETE, 0,
-				   op == OP_UNDELETE_THREAD ? 0 : 1);
+				   op == OP_UNDELETE_THREAD ? 0 : 1)
+	  + mutt_thread_set_flag (CURHDR, M_PURGED, 0,
+				  op == OP_UNDELETE_THREAD ? 0 : 1);
 
-	if (rc != -1)
+	if (rc > -1)
 	{
diff -pruN1 mutt-1.3.25.trash_folder/flags.c mutt-1.3.25.purge_command/flags.c
--- mutt-1.3.25.trash_folder/flags.c	Wed Jan  9 20:39:31 2002
+++ mutt-1.3.25.purge_command/flags.c	Wed Jan  9 20:40:27 2002
@@ -100,2 +100,12 @@ void _mutt_set_flag (CONTEXT *ctx, HEADE
 
+    case M_PURGED:
+      if (bf)
+      {
+	if (!h->purged)
+	  h->purged = 1;
+      }
+      else if (h->purged)
+	h->purged = 0;
+      break;
+
     case M_NEW:
diff -pruN1 mutt-1.3.25.trash_folder/functions.h mutt-1.3.25.purge_command/functions.h
--- mutt-1.3.25.trash_folder/functions.h	Tue Sep 11 12:51:39 2001
+++ mutt-1.3.25.purge_command/functions.h	Wed Jan  9 20:40:27 2002
@@ -100,2 +100,3 @@ struct binding_t OpMain[] = {
   { "next-subthread",		OP_MAIN_NEXT_SUBTHREAD,		"\033n" },
+  { "purge-message",		OP_PURGE_MESSAGE,		NULL },
   { "query",			OP_QUERY,			"Q" },
@@ -184,2 +185,3 @@ struct binding_t OpPager[] = {
   { "previous-subthread",OP_MAIN_PREV_SUBTHREAD,	"\033p" },
+  { "purge-message",	OP_PURGE_MESSAGE,		NULL },
   { "quit",		OP_QUIT,			"Q" },
diff -pruN1 mutt-1.3.25.trash_folder/mutt.h mutt-1.3.25.purge_command/mutt.h
--- mutt-1.3.25.trash_folder/mutt.h	Wed Jan  9 20:39:31 2002
+++ mutt-1.3.25.purge_command/mutt.h	Wed Jan  9 20:40:27 2002
@@ -187,2 +187,3 @@ enum
   M_APPENDED,
+  M_PURGED,
   M_FLAG,
@@ -620,2 +621,3 @@ typedef struct header
   unsigned int appended : 1; /* has been saved */
+  unsigned int purged : 1;   /* bypassing the trash folder */
   unsigned int deleted : 1;
diff -pruN1 mutt-1.3.25.trash_folder/mx.c mutt-1.3.25.purge_command/mx.c
--- mutt-1.3.25.trash_folder/mx.c	Wed Jan  9 20:39:31 2002
+++ mutt-1.3.25.purge_command/mx.c	Wed Jan  9 20:40:27 2002
@@ -841,2 +841,3 @@ static int trash_append (CONTEXT *ctx)
 	if (ctx->hdrs[i]->deleted && !ctx->hdrs[i]->appended
+	    && !ctx->hdrs[i]->purged
 	    && mutt_append_message (ctx_trash, ctx, ctx->hdrs[i], 0, 0) == -1)
diff -pruN1 mutt-1.3.25.trash_folder/pager.c mutt-1.3.25.purge_command/pager.c
--- mutt-1.3.25.trash_folder/pager.c	Fri Dec 21 01:21:03 2001
+++ mutt-1.3.25.purge_command/pager.c	Wed Jan  9 20:40:27 2002
@@ -2148,2 +2148,3 @@ mutt_pager (const char *banner, const ch
 
+      case OP_PURGE_MESSAGE:
       case OP_DELETE:
@@ -2152,2 +2153,3 @@ mutt_pager (const char *banner, const ch
 	mutt_set_flag (Context, extra->hdr, M_DELETE, 1);
+	mutt_set_flag (Context, extra->hdr, M_PURGED, 1);
         if (option (OPTDELETEUNTAG))
@@ -2461,2 +2463,3 @@ mutt_pager (const char *banner, const ch
 	mutt_set_flag (Context, extra->hdr, M_DELETE, 0);
+	mutt_set_flag (Context, extra->hdr, M_PURGED, 0);
 	redraw = REDRAW_STATUS | REDRAW_INDEX;
@@ -2475,5 +2478,7 @@ mutt_pager (const char *banner, const ch
 	r = mutt_thread_set_flag (extra->hdr, M_DELETE, 0,
+				  ch == OP_UNDELETE_THREAD ? 0 : 1)
+	  + mutt_thread_set_flag (extra->hdr, M_PURGED, 0,
 				  ch == OP_UNDELETE_THREAD ? 0 : 1);
 
-	if (r != -1)
+	if (r > -1)
 	{
diff -pruN1 mutt-1.3.25.trash_folder/pattern.c mutt-1.3.25.purge_command/pattern.c
--- mutt-1.3.25.trash_folder/pattern.c	Mon Dec  3 11:17:57 2001
+++ mutt-1.3.25.purge_command/pattern.c	Wed Jan  9 20:40:27 2002
@@ -1172,4 +1172,6 @@ int mutt_pattern_func (int op, char *pro
 	{
-	  case M_DELETE:
 	  case M_UNDELETE:
+	    mutt_set_flag (Context, Context->hdrs[Context->v2r[i]], M_PURGED,
+			   0);
+	  case M_DELETE:
 	    mutt_set_flag (Context, Context->hdrs[Context->v2r[i]], M_DELETE, 
diff -pruN mutt-1.3.25.orig/PATCHES mutt-1.3.25/PATCHES
--- mutt-1.3.25.orig/PATCHES    Mon Nov 26 20:16:52 2001
+++ mutt-1.3.25/PATCHES Thu Dec  6 16:27:55 2001
@@ -1,0 +1 @@
+patch-1.3.25.cd.purge_command.2
