--- mutt-1.4.2.1/Makefile.in.orig	2004-02-12 09:52:26.000000000 +0100
+++ mutt-1.4.2.1/Makefile.in	2004-02-14 11:14:52.096777216 +0100
@@ -117,7 +117,7 @@
 BUILT_SOURCES = keymap_defs.h patchlist.c
 
 bin_PROGRAMS = mutt @DOTLOCK_TARGET@ @PGPAUX_TARGET@
-mutt_SOURCES = $(BUILT_SOURCES) 	addrbook.c alias.c attach.c base64.c browser.c buffy.c color.c 	commands.c complete.c compose.c copy.c curs_lib.c curs_main.c date.c 	edit.c enter.c flags.c init.c filter.c from.c getdomain.c 	handler.c hash.c hdrline.c headers.c help.c hook.c keymap.c 	main.c mbox.c menu.c mh.c mx.c pager.c parse.c pattern.c 	postpone.c query.c recvattach.c recvcmd.c 	rfc822.c rfc1524.c rfc2047.c rfc2231.c 	score.c send.c sendlib.c signal.c sort.c 	status.c system.c thread.c charset.c history.c lib.c 	muttlib.c editmsg.c utf8.c mbyte.c wcwidth.c 	url.c ascii.c
+mutt_SOURCES = $(BUILT_SOURCES) 	addrbook.c alias.c attach.c base64.c browser.c buffy.c color.c 	commands.c complete.c compose.c compress.c copy.c curs_lib.c curs_main.c date.c 	edit.c enter.c flags.c init.c filter.c from.c getdomain.c 	handler.c hash.c hdrline.c headers.c help.c hook.c keymap.c 	main.c mbox.c menu.c mh.c mx.c pager.c parse.c pattern.c 	postpone.c query.c recvattach.c recvcmd.c 	rfc822.c rfc1524.c rfc2047.c rfc2231.c 	score.c send.c sendlib.c signal.c sort.c 	status.c system.c thread.c charset.c history.c lib.c 	muttlib.c editmsg.c utf8.c mbyte.c wcwidth.c 	url.c ascii.c
 
 
 mutt_LDADD = @MUTT_LIB_OBJECTS@ @LIBOBJS@ $(LIBIMAP) $(MUTTLIBS) 	$(INTLLIBS) $(LIBICONV)
@@ -151,9 +151,8 @@
 
 EXTRA_mutt_SOURCES = account.c md5c.c mutt_sasl.c mutt_socket.c mutt_ssl.c 	mutt_tunnel.c pop.c pop_auth.c pop_lib.c pgp.c pgpinvoke.c pgpkey.c 	pgplib.c sha1.c pgpmicalg.c gnupgparse.c resize.c dotlock.c remailer.c 	browser.h mbyte.h remailer.h url.h mutt_ssl_nss.c pgppacket.c 
 
-
-EXTRA_DIST = COPYRIGHT GPL OPS OPS.PGP TODO configure acconfig.h account.h 	attach.h buffy.h charset.h copy.h dotlock.h functions.h gen_defs 	globals.h hash.h history.h init.h keymap.h 	mailbox.h mapping.h md5.h mime.h mutt.h mutt_curses.h mutt_menu.h 	mutt_regex.h mutt_sasl.h mutt_socket.h mutt_ssl.h mutt_tunnel.h 	mx.h pager.h pgp.h pop.h protos.h reldate.h rfc1524.h rfc2047.h 	rfc2231.h rfc822.h sha1.h sort.h mime.types VERSION prepare 	_regex.h OPS.MIX README.SECURITY remailer.c remailer.h browser.h 	mbyte.h	lib.h extlib.c pgpewrap.c pgplib.h Muttrc.head Muttrc 	makedoc.c stamp-doc-rc README.SSL 	muttbug pgppacket.h depcomp ascii.h BEWARE PATCHES patchlist.sh 	snprintf.c regex.c
-
+EXTRA_DIST = COPYRIGHT GPL OPS OPS.PGP TODO configure acconfig.h account.h 	attach.h buffy.h charset.h compress.h copy.h dotlock.h functions.h gen_defs 	globals.h hash.h history.h init.h keymap.h 	mailbox.h mapping.h md5.h mime.h mutt.h mutt_curses.h mutt_menu.h 	mutt_regex.h mutt_sasl.h mutt_socket.h mutt_ssl.h mutt_tunnel.h 	mx.h pager.h pgp.h pop.h protos.h reldate.h rfc1524.h rfc2047.h 	rfc2231.h rfc822.h sha1.h sort.h mime.types VERSION prepare 	_regex.h OPS.MIX README.SECURITY remailer.c remailer.h browser.h 	mbyte.h	lib.h extlib.c pgpewrap.c pgplib.h Muttrc.head Muttrc 	makedoc.c stamp-doc-rc README.SSL 	muttbug pgppacket.h depcomp ascii.h BEWARE PATCHES patchlist.sh
+ 
 
 mutt_dotlock_SOURCES = mutt_dotlock.c
 mutt_dotlock_LDADD = @LIBOBJS@
@@ -187,7 +186,7 @@
 makedoc_OBJECTS =  makedoc.o
 makedoc_LDFLAGS = 
 mutt_OBJECTS =  patchlist.o addrbook.o alias.o attach.o base64.o \
-browser.o buffy.o color.o commands.o complete.o compose.o copy.o \
+browser.o buffy.o color.o commands.o complete.o compose.o compress.o copy.o \
 curs_lib.o curs_main.o date.o edit.o enter.o flags.o init.o filter.o \
 from.o getdomain.o handler.o hash.o hdrline.o headers.o help.o hook.o \
 keymap.o main.o mbox.o menu.o mh.o mx.o pager.o parse.o pattern.o \
