--- mutt-2.2.13/doc/Muttrc.orig	2024-10-27 16:52:16.897389301 +0100
+++ mutt-2.2.13/doc/Muttrc	2024-10-27 17:10:05.321601153 +0100
@@ -1804,6 +1804,7 @@ attachments   -I message/external-body
 # 
 # 
 # set forward_format="[%a: %s]"
+set forward_format="Fwd: %s"
 #
 # Name: forward_format
 # Type: string
@@ -2036,6 +2037,7 @@ attachments   -I message/external-body
 # 
 # 
 # set history=10
+set history=100
 #
 # Name: history
 # Type: number
@@ -2846,6 +2848,7 @@ attachments   -I message/external-body
 # 
 # 
 # set mark_old=yes
+set mark_old=no
 #
 # Name: mark_old
 # Type: boolean
@@ -3101,6 +3104,7 @@ attachments   -I message/external-body
 # 
 # 
 # set mime_forward=no
+set mime_forward=ask-no
 #
 # Name: mime_forward
 # Type: quadoption
@@ -3298,6 +3302,7 @@ attachments   -I message/external-body
 # 
 # 
 # set pager_context=0
+set pager_context=1
 #
 # Name: pager_context
 # Type: number
@@ -3328,6 +3333,7 @@ attachments   -I message/external-body
 # 
 # 
 # set pager_index_lines=0
+set pager_index_lines=6
 #
 # Name: pager_index_lines
 # Type: number
@@ -4351,6 +4357,7 @@ attachments   -I message/external-body
 # 
 # 
 # set quit=yes
+set quit=ask-yes
 #
 # Name: quit
 # Type: quadoption
@@ -4537,6 +4544,7 @@ attachments   -I message/external-body
 # 
 # 
 # set reply_regexp="^(re)(\\\\[[0-9]+\\\\])*:[ \t]*"
+set reply_regexp="^(([oO][dD][pP]|[aA][wW]|[rR][eE]([\\[0-9\\]+])*:|[Oo][dD][pP]:|[aA][wW]:|[rR][eE]:)[ \t]+)*"
 #
 # Name: reply_regexp
 # Type: regular expression (localized)
@@ -4853,6 +4861,7 @@ attachments   -I message/external-body
 # 
 # 
 # set send_charset="us-ascii:iso-8859-1:utf-8"
+set send_charset="us-ascii:utf-8"
 #
 # Name: send_charset
 # Type: string
@@ -5231,6 +5240,7 @@ attachments   -I message/external-body
 # 
 # 
 # set simple_search="~f %s | ~s %s"
+set simple_search="~f %s | ~C %s | ~s %s"
 #
 # Name: simple_search
 # Type: string
@@ -5748,6 +5758,7 @@ attachments   -I message/external-body
 # 
 # 
 # set sort=date
+set sort=threads
 #
 # Name: sort
 # Type: sort order
@@ -6341,6 +6352,7 @@ attachments   -I message/external-body
 # 
 # 
 # set thorough_search=yes
+set thorough_search=yes
 #
 # Name: thorough_search
 # Type: boolean
@@ -6371,6 +6383,7 @@ attachments   -I message/external-body
 # 
 # 
 # set tilde=no
+set tilde=yes
 #
 # Name: tilde
 # Type: boolean
@@ -6780,3 +6793,4 @@ attachments   -I message/external-body
 # ``tuning'' section of the manual for performance considerations.
 # 
 # 
+source mutt_source-muttrc.d|
