--- mutt-1.5.21/configure.ac~	2010-10-06 19:50:24.570430998 +0200
+++ mutt-1.5.21/configure.ac	2010-10-06 20:17:25.192640340 +0200
@@ -270,12 +270,13 @@
 
 	AC_CHECK_FUNC(initscr,,[
 	cf_ncurses="ncurses"
-	for lib in ncurses ncursesw
+	cf_tinfo="tinfo"
+	for lib in "" w
 	do
-		AC_CHECK_LIB($lib, waddnwstr, [cf_ncurses="$lib"; break])
+		AC_CHECK_LIB(ncurses$lib, waddnwstr, [cf_ncurses="ncurses$lib"; cf_tinfo="tinfo$lib"; break])
 	done
         AC_CHECK_LIB($cf_ncurses, initscr,
-                [MUTTLIBS="$MUTTLIBS -l$cf_ncurses"
+                [MUTTLIBS="$MUTTLIBS -l$cf_ncurses -l$cf_tinfo"
 		
                 if test "$cf_ncurses" = ncursesw; then
 			AC_CHECK_HEADERS(ncursesw/ncurses.h,[cf_cv_ncurses_header="ncursesw/ncurses.h"])
