diff -urNp mutt-1.5.20.orig/sendlib.c mutt-1.5.20/sendlib.c
--- mutt-1.5.20.orig/sendlib.c	2009-06-14 17:46:11.000000000 +0200
+++ mutt-1.5.20/sendlib.c	2010-09-11 11:49:12.000000000 +0200
@@ -1737,7 +1737,7 @@ static int write_one_header (FILE *fp, i
   }
   else
   {
-    t = strchr (start, ':');
+    if(strncmp("From ",start,5)==0) t = strchr (start, ' '); else t = strchr (start, ':');
     tagbuf = mutt_substrdup (start, t);
     valbuf = mutt_substrdup (t + 2, end);
     dprint(4,(debugfile,"mwoh: buf[%s%s] too long, "
