--- mutt-1.5.4/configure.in.orig	Thu Mar 20 17:34:41 2003
+++ mutt-1.5.4/configure.in	Thu Mar 20 17:43:44 2003
@@ -339,8 +339,7 @@
 AC_CHECK_FUNCS(snprintf, , [mutt_cv_snprintf=yes])
 AC_CHECK_FUNCS(vsnprintf, , [mutt_cv_snprintf=yes])
 if test $mutt_cv_snprintf = yes; then
-#        AC_LIBOBJ(snprintf)
-        LIBOBJS="$LIBOBJS snprintf.o"
+        AC_LIBOBJ(snprintf)
 fi
 
 dnl SCO uses chsize() instead of ftruncate()
@@ -372,8 +371,7 @@
 
 if test $mutt_cv_regex = yes; then
         AC_DEFINE(USE_GNU_REGEX,1,[ Define if you want to use the included regex.c. ])
-###     AC_LIBOBJ(regex)
-        LIBOBJS="$LIBOBJS regex.o"
+	AC_LIBOBJ(regex)
 fi
 
 
